### PR TITLE
Commodities Endpoint Integration

### PR DESCRIPTION
**Pull request by ericdudley**
Writing the API integration code for the `/commodities` endpoint.

Related [backend issue](https://github.com/agblox/research-lab/issues/49).

### CODE REVIEW

**Comment by tbosch82** [comment]
@ericdudley Take a look. The functionality to dynamically pull the available products and contracts is currently just in the trading screen (just to be sure it is working correctly). I will open another issue to pull everything together into the individual charts.

File: demo/src/Dashboard/Chart/Chart.tsx [hunk 1]
```diff
@@ -356,7 +356,9 @@ class Chart extends Component<ChartProps & StoreProps> {
                           <SubTitle>
                             {stat.name}{" "}
                             {activeDashboard!.commodity.contract &&
-                              activeDashboard!.commodity.contract.label}
+                              activeDashboard!.commodity.contract.label.split(
+                                "("
```
Comment (ericdudley) [inline demo/src/Dashboard/Chart/Chart.tsx]: I'm not sure what this is doing.

File: demo/src/Dashboard/Dashboard.tsx [hunk 2]
```diff
@@ -102,6 +97,7 @@ class Dashboard extends Component<
     } else if (!id_regex && selectedDashboard) {
       push(`/dashboard/${selectedDashboard.value}`);
     }
+    this.props.store!.loadProducts();
```
Comment (ericdudley) [inline demo/src/Dashboard/Dashboard.tsx]: Can add `loadProducts` to line 88.

File: demo/src/Trading/TradingOverview/TradingOverview.tsx [hunk 3]
```diff
@@ -29,29 +29,44 @@ class TradingOverview extends Component<StoreProps & TradingOverviewProps> {
         {comm ? (
           <div className="trading-overview">
             <SubTitle className="trading-overview__header">
-              {commodity.name!.label}{" "}
-              {commodity.contract && commodity.contract.label}
+              {commodity.name!.label.split("(")[0]}{" "}
+              {commodity.contract && commodity.contract.label.split("(")[0]}
```
Comment (ericdudley) [inline demo/src/Trading/TradingOverview/TradingOverview.tsx]: Same as above, seems pretty weak.

File: demo/src/Trading/TradingOverview/TradingOverview.tsx [hunk 4]
```diff
@@ -29,29 +29,44 @@ class TradingOverview extends Component<StoreProps & TradingOverviewProps> {
...
-                {comm.last}
-              </li>
-              <li>
-                <span className="trading-overview__stats__option">Close</span>
-                {comm.close}
+                <span className="trading-overview__stats__option">Open</span>
+                {comm.open &&
+                  (comm.open[comm.open.length - 1]
+                    ? comm.open[comm.open.length - 1].toFixed(2)
+                    : "No Data")}
```
Comment (ericdudley) [inline demo/src/Trading/TradingOverview/TradingOverview.tsx]: Add this to the `copy` module.

File: demo/src/util/api.ts [hunk 5]
```diff
@@ -39,8 +39,20 @@ export const DIESEL_DATA = {
 
 export const TWEET_DATA = sdata1;
 
+export function getProducts() {
+  return axios
+    .get(`${api_base}/commodities/products?`, axios_config())
```
Comment (ericdudley) [inline demo/src/util/api.ts]: Doesn't need the `?`.

File: demo/src/util/api.ts [hunk 6]
```diff
@@ -39,8 +39,20 @@ export const DIESEL_DATA = {
...
+    .then(resp => {
+      if (resp.data) {
+        let data;
+        data = resp.data;
+        return data;
```
Comment (ericdudley) [inline demo/src/util/api.ts]: I probably did this somewhere, but this can just be `return resp.data;`.

File: demo/src/util/api.ts [hunk 7]
```diff
@@ -57,6 +69,44 @@ export function getData(source: Source) {
...
         }
       });
+  } else if (
+    [SourceIds.FEEDERCATTLE, SourceIds.CORN, SourceIds.GASOLINE].includes(
+      id as SourceIds
+    )
+  ) {
+    return axios
+      .get(
+        `${api_base}/commodities?start=${tsToDateString(
```
Comment (ericdudley) [inline demo/src/util/api.ts]: There might be a much cleaner way to do the query variables with axios. Something like a `query` object.

File: demo/src/util/api.ts [hunk 8]
```diff
@@ -57,6 +69,44 @@ export function getData(source: Source) {
...
+          source.startTime
+        )}&end=${tsToDateString(source.endTime)}&product_code=${id}${
+          contract ? "&expiration_code=" + contract.value : ""
+        }`,
+        axios_config()
+      )
+      .then(resp => {
+        if (resp.data) {
+          /*resp.data.sort(function(a, b) {
```
Comment (ericdudley) [inline demo/src/util/api.ts]: Remove commented out code.

File: demo/src/util/figure.ts [hunk 9]
```diff
@@ -294,22 +293,25 @@ export class FigureStore {
     const figure = (this.figureCache.get(id) as FigureCacheEntry).figure;
     figure.stats = figure.data.map(({ x, y, name }) => ({
       id: getRandomId(),
-      name: name,
+      name: name.split("(")[0],
```
Comment (ericdudley) [inline demo/src/util/figure.ts]: Again

File: demo/src/util/store.ts [hunk 10]
```diff
@@ -1,10 +1,11 @@
 import { observable, action, computed, reaction } from "mobx";
 import copy from "./copy";
-import { loadConfig, saveConfig } from "./api";
+import { loadConfig, saveConfig } from "./local_storage";
```
Comment (ericdudley) [inline demo/src/util/store.ts]: Switch back to api.

File: demo/src/util/store.ts [hunk 11]
```diff
@@ -117,32 +119,30 @@ export enum TradingActionLabels {
...
-  JUL19 = "Jul 2019",
-  SEP19 = "Sep 2019",
-  DEC19 = "Dec 2019",
-  MAR20 = "Mar 2019",
-  MAY20 = "May 2020",
-  JUL20 = "Jul 2020",
-  SEP20 = "Sep 2020",
-  DEC20 = "Dec 2020"
+export enum GasolineContractIds {
+  N9 = "Jul 2019",
```
Comment (ericdudley) [inline demo/src/util/store.ts]: I would write a util function for this. All the products have the same form.
https://www.cmegroup.com/month-codes.html

File: demo/src/util/data_loader.ts [hunk 12]
```diff
@@ -1,15 +1,21 @@
...
+  `${source.id}|||${source.startTime}|||${source.endTime}|||${
+    source.keywords
+  }|||${source.contract && source.contract!.value}`;
 
 const MAX_ATTEMPTS = 3;
 const data: Map<string, any> = new Map();
 const dataAttempts: Map<string, any> = new Map();
 const dataLoading: Map<string, Promise<any>> = new Map();
 
+export function getCommInfo() {
```
Comment (ericdudley) [inline demo/src/util/data_loader.ts]: You can probably just get rid of this function if data_loader isn't actually doing anything.

File: demo/src/util/store.ts [hunk 13]
```diff
@@ -731,12 +734,73 @@ class Store {
...
+      this.activeDashboard.commodity.name
+    ) {
+      this.loadContracts();
+      return this.contracts.get(this.activeDashboard.commodity.name.value);
+    }
+    return [];
+  }
+
+  loadContracts() {
+    getCommInfo().then(_data => {
```
Comment (ericdudley) [inline demo/src/util/store.ts]: Is there a reason why this and `loadProducts` is separate.

File: demo/src/util/store.ts [hunk 14]
```diff
@@ -731,12 +734,73 @@ class Store {
...
+  }
+
+  @computed
+  public get getContracts() {
+    if (
+      this.activeDashboard &&
+      this.products &&
+      this.activeDashboard.commodity.name
+    ) {
+      this.loadContracts();
```
Comment (ericdudley) [inline demo/src/util/store.ts]: Loading the products/contracts from the API should probably not happen every time we need the contracts in the frontend.

File: demo/src/Dashboard/Chart/Chart.tsx (hunk 1 above)
Comment (tbosch82) [reply demo/src/Dashboard/Chart/Chart.tsx]: This isn't necessary anymore. It was to split the way the names were hard coded before.

File: demo/src/Trading/TradingOverview/TradingOverview.tsx (hunk 3 above)
Comment (tbosch82) [reply demo/src/Trading/TradingOverview/TradingOverview.tsx]: Same as above.

File: demo/src/Trading/TradingOverview/TradingOverview.tsx (hunk 4 above)
Comment (tbosch82) [reply demo/src/Trading/TradingOverview/TradingOverview.tsx]: @ericdudley You just want "No Data" put into the copy module?

Comment (ericdudley) [reply demo/src/Trading/TradingOverview/TradingOverview.tsx]: Yeah, so we can have human readable stuff mostly in one place.

**Comment by tbosch82** [comment]
@ericdudley Okay, I think I hit all your comments, let me know what you think. My goal now is to integrate all of this into the individual charts (which will be a pain since every commodity offers different contracts, so I figured that would be best tackled in another issue/PR). Once that is done the store.ts code should be cleaned up nicely because a lot of the getters and enums that are producing hard-coded options will be able to be removed.

File: demo/src/util/api.ts [hunk 15]
```diff
@@ -27,6 +27,29 @@ function axios_config() {
...
 
+interface Query {
+  product_code?: string;
+  start?: string;
+  end?: string;
+  text?: string;
+  expiration_code?: string;
+}
+
+function buildQuery(data?: Query) {
```
Comment (ericdudley) [inline demo/src/util/api.ts]: What I meant for the query params was to use axios. It looks like axios has an example in their docs.
https://www.npmjs.com/package/axios
Search for `// Optionally the request above could also be done as` in the page.

File: demo/src/util/util.ts [hunk 16]
```diff
@@ -7,6 +7,38 @@ export interface RoutingProps {
...
+enum ContractYears {
+  _9 = "19",
+  _0 = "20",
+  _1 = "21",
+  _2 = "22"
+}
+
+export function getContractIds(x: string): string {
+  const month = Object.entries(ContractMonths).find(
+    e => e[0] === x.split("")[0]
```
Comment (ericdudley) [inline demo/src/util/util.ts]: nit: I think you're looking for `.charAt(0)`?

File: demo/src/util/util.ts [hunk 17]
```diff
@@ -7,6 +7,38 @@ export interface RoutingProps {
...
+}
+
+enum ContractYears {
+  _9 = "19",
+  _0 = "20",
+  _1 = "21",
+  _2 = "22"
+}
+
+export function getContractIds(x: string): string {
```
Comment (ericdudley) [inline demo/src/util/util.ts]: nit: The function signature doesn't describe very well what this does.

**Comment by ericdudley** [comment]
@tbosch82 Left  few more comments, after those feel free to merge.

**Pull request merged**